cmd/dra-example-controller/gpu.go

```diff
@@ -0,0 +1,152 @@
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	resourcev1 "k8s.io/api/resource/v1alpha2"
+	"k8s.io/dynamic-resource-allocation/controller"
+
+	nascrd "github.com/kubernetes-sigs/dra-example-driver/api/example.com/resource/gpu/nas/v1alpha1"
+	gpucrd "github.com/kubernetes-sigs/dra-example-driver/api/example.com/resource/gpu/v1alpha1"
+)
+
+type gpudriver struct {
+	PendingClaimRequests *PerNodeClaimRequests
+}
+
+func NewGpuDriver() *gpudriver {
+	return &gpudriver{
+		PendingClaimRequests: NewPerNodeClaimRequests(),
+	}
+}
+
+func (g *gpudriver) ValidateClaimParameters(claimParams *gpucrd.GpuClaimParametersSpec) error {
+	if claimParams.Count < 1 {
+		return fmt.Errorf("invalid number of GPUs requested: %v", claimParams.Count)
+	}
+	return nil
+}
+
+func (g *gpudriver) Allocate(crd *nascrd.NodeAllocationState, claim *resourcev1.ResourceClaim, claimParams *gpucrd.GpuClaimParametersSpec, class *resourcev1.ResourceClass, classParams *gpucrd.DeviceClassParametersSpec, selectedNode string) (OnSuccessCallback, error) {
+	claimUID := string(claim.UID)
+
+	if !g.PendingClaimRequests.Exists(claimUID, selectedNode) {
+		return nil, fmt.Errorf("no allocation requests generated for claim '%v' on node '%v' yet", claim.UID, selectedNode)
+	}
+
+	crd.Spec.ClaimRequests[claimUID] = g.PendingClaimRequests.Get(claimUID, selectedNode)
+	onSuccess := func() {
+		g.PendingClaimRequests.Remove(claimUID)
+	}
+
+	return onSuccess, nil
+}
+
+func (g *gpudriver) Deallocate(crd *nascrd.NodeAllocationState, claim *resourcev1.ResourceClaim) error {
+	g.PendingClaimRequests.Remove(string(claim.UID))
+	return nil
+}
+
+func (g *gpudriver) UnsuitableNode(crd *nascrd.NodeAllocationState, pod *corev1.Pod, gpucas []*controller.ClaimAllocation, allcas []*controller.ClaimAllocation, potentialNode string) error {
+	g.PendingClaimRequests.VisitNode(potentialNode, func(claimUID string, request nascrd.RequestedDevices) {
+		if _, exists := crd.Spec.ClaimRequests[claimUID]; exists {
+			g.PendingClaimRequests.Remove(claimUID)
+		} else {
+			crd.Spec.ClaimRequests[claimUID] = request
+		}
+	})
+
+	allocated := g.allocate(crd, pod, gpucas, allcas, potentialNode)
+	for _, ca := range gpucas {
+		claimUID := string(ca.Claim.UID)
+		claimParams := ca.ClaimParameters.(*gpucrd.GpuClaimParametersSpec)
+
+		if claimParams.Count != len(allocated[claimUID]) {
+			for _, ca := range allcas {
+				ca.UnsuitableNodes = append(ca.UnsuitableNodes, potentialNode)
+			}
+			return nil
+		}
+
+		var devices []nascrd.RequestedGpu
+		for _, gpu := range allocated[claimUID] {
+			device := nascrd.RequestedGpu{
+				UUID: gpu,
+			}
+			devices = append(devices, device)
+		}
+
+		requestedDevices := nascrd.RequestedDevices{
+			Gpu: &nascrd.RequestedGpus{
+				Devices: devices,
+			},
+		}
+
+		g.PendingClaimRequests.Set(claimUID, potentialNode, requestedDevices)
+		crd.Spec.ClaimRequests[claimUID] = requestedDevices
```

Comment:
Why do we need this line of code even though the ClaimRequests (AllocatedClaims) field of the NAS is updated in the Allocate API?
https://github.com/kubernetes-sigs/dra-example-driver/blob/cba2ec61cfedf51600d678f340de5c5f2d847fab/cmd/dra-example-controller/gpu.go#L54

Reply:
it's not needed, you're right, CRD is not updated when UnsuitableNodes is called, there is no decision at that point, only when the Allocate is called.

Reply:
It is not needed in this example driver. This is leftover after porting this function from the original code in the NVIDIA GPU driver, where it _is_ needed due to this loop:
https://github.com/NVIDIA/k8s-dra-driver/blob/main/cmd/nvidia-dra-controller/driver.go#L281-L292

Reply:
Ah, I see. I was wondering why that was there, and it helped me to know. Thanks for the explanation!